Device setup

Spectre5 opened the pull request:
## Description: Add support to reset and setup pywemo devices

This is a subset of the closed PR #178 where only the essential reset and setup methods are created for the Device class.

**Related issue (if applicable):** fixes #171, #161

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] There is no commented out code in this PR.

Spectre5 committed:
ignore *.code-workspace

Spectre5 committed:
initial reset and setup method implementations

Spectre5 committed:
add return values

Spectre5 committed:
add firmware warning to README

Spectre5 commented:
@pavoni have you had any time to take a look at this yet?  I'm happy to look at any inputs you have, style or otherwise.

pavoni commented:
It’s been a crazy week. I had a quick look and it looked great - but wanted to take a closer look and try the code.

Will get to it this week!

Spectre5 commented:
Ok, no rush!  Mostly I just wanted to make sure you didn't miss the notification of the new PR

Spectre5 committed:
catch UnknownService in setup method

Spectre5 committed:
factory reset method should return the status

Spectre5 committed:
remove -salt to openssl; -S <salt> implies it

Spectre5 committed:
minor clean up

pavoni commented on pywemo/ouimeaux_device/__init__.py:
```diff
@@ -236,6 +256,293 @@ def explain(self):
+
+        # call OpenSSL to encrypt the data
+        try:
+            openssl = subprocess.run(
```
This will require `openssl` installed on the machine - which is fine (I think practically most people probably will) - but this probably wouldn't work on a standard windows machine.

Let's just add a note to the README to say this is require to use the setup methods.

pavoni commented on pywemo/ouimeaux_device/__init__.py:
```diff
@@ -236,6 +256,293 @@ def explain(self):
+
+        From testing on a handful of devices, the Reset codes used in the
+        ReSetup action below were consistent.  These could potentially change
+        in the future or may be different for other untested devices.
```
Could you just add a couple of words on what the two parameters do?

pavoni commented on pywemo/ouimeaux_device/__init__.py:
```diff
@@ -236,6 +256,293 @@ def explain(self):
+
+    def setup(self, *args, **kwargs):
+        """Interface method for device setup."""
+        try:
```
Again -perhaps a couple of words on the parameters and what they do.

pavoni reviewed (commented):
This is really great.

It worked perfectly on my insight switch. It's actually more reliable than the WeMo app!

It sort of works on my Wemo Maker - although gave an error after sending the network info. The WeMo Maker connects fine - so it's just the follow up status info that seems not to.

I'll do a bit more digging to see if I can see what's happening.

pavoni commented on pywemo/ouimeaux_device/__init__.py:
```diff
@@ -236,6 +256,293 @@ def explain(self):
+            status = None
+            while time.time() - timeout_start < timeout and status != '1':
+                time.sleep(1.5)
+                status = wifisetup.GetNetworkStatus()['NetworkStatus']
```
So it looks like the WeMoMaker has already disconnected at this point - and even if I remove the sleep - it has already disconnected. 

So perhaps wrap a try block around this and handle the fallout?

Can make a suggestion if that would be helpful.

In my testing BTW the code always worked, just reported failure!

Spectre5 commented:
Awesome, thanks for testing it out.  I'll address your comments soon!

As for the Maker, it sounds like it connects and closes its local access point within the first 1.5 second delay (plus the 0.10 second delay between the two `ConnectHomeNetwork` commands).  Is that what happens?  I had considered that possibility, but none of my devices ever closed that quickly.  Ideally the devices would not close their local access point until the `CloseSetup` method was called!  But not all of my devices even have that method.  Possibly Belkin normally calls that method after the device connects to the home network, I'm not really sure.

Can you include the error message and/or stack trace that you get with the Maker?  Also, can you show the entire printed result from the `explain` method?  I'm curious to see if there are any methods on it that are not on my devices that sound like they could be related.

Spectre5 commented:
Ah wait, I just noticed that you said even if you remove the sleep that the Maker already disconnected.  Was the error identical when you removed the sleep?  If you run it with the debug log on, do you see any of those `network status after X seconds` messages when there is no sleep or only a very short one?

What if you change `for _ in range(2):` to just `for _ in range(1)` so that it only sends 1 `ConnectHomeNetwork` command for the Maker.  Does that make a difference?

Perhaps the firmware on the Maker just disconnects very quickly after `ConnectHomeNetwork` (although it does seem that you get 2 of those sent since you don't get an error there) and then the Wemo app reconnects to the home network and waits to see if the Maker joins it....just a possibility.

pavoni commented:
So I've experimented a but - reducing the `range`to `1` just makes it unreliable.

I'm now back to your code - which gives an error / reconnect attempt - but seems to work.

Logs here
```
>>> devices = pywemo.discover_devices()
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setup.xml HTTP/1.1" 200 4882
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setup.xml HTTP/1.1" 200 4882
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setup.xml HTTP/1.1" 200 4882
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setupservice.xml HTTP/1.1" 200 3278
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /timesyncservice.xml HTTP/1.1" 200 2190
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /eventservice.xml HTTP/1.1" 200 16357
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /deviceservice.xml HTTP/1.1" 200 3075
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /firmwareupdate.xml HTTP/1.1" 200 1826
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /rulesservice.xml HTTP/1.1" 200 11218
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /metainfoservice.xml HTTP/1.1" 200 1132
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /remoteaccess.xml HTTP/1.1" 200 3974
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /deviceinfoservice.xml HTTP/1.1" 200 10704
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /smartsetup.xml HTTP/1.1" 200 2539
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /manufacture.xml HTTP/1.1" 200 642
>>> devices
[<WeMo Maker "Hi Fi Systemline Sensor">]
>>> devices[0].setup("Elsden-h", "PASSWORD REMOVED")
INFO:pywemo.ouimeaux_device:scanning for AP's...
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "POST /upnp/control/WiFiSetup1 HTTP/1.1" 200 331
DEBUG:pywemo.ouimeaux_device:found AP: Elsden-h|11|100|WPA2PSK/AES
INFO:pywemo.ouimeaux_device:selecting AP: Elsden-h|11|100|WPA2PSK/AES
DEBUG:pywemo.ouimeaux_device:AP channel: 11
DEBUG:pywemo.ouimeaux_device:AP authorization mode(s): WPA2PSK
DEBUG:pywemo.ouimeaux_device:AP encryption method: AES
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "POST /upnp/control/metainfo1 HTTP/1.1" 200 366
DEBUG:pywemo.ouimeaux_device:password length (before encryption): 8
DEBUG:pywemo.ouimeaux_device:password length (after encryption): 24
INFO:pywemo.ouimeaux_device:sending connection request (try 1)
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "POST /upnp/control/WiFiSetup1 HTTP/1.1" 200 305
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "POST /upnp/control/WiFiSetup1 HTTP/1.1" 200 305
INFO:pywemo.ouimeaux_device:pairing status: {'PairingStatus': 'Connecting'}
INFO:pywemo.ouimeaux_device:starting status checks (20 second timeout)
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
WARNING:pywemo.ouimeaux_device.api.service:Error communicating with Hi Fi Systemline Sensor at 10.22.22.1:49152, retry 0
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49153
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49154
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49151
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49155
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49156
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49157
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49158
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49159
ERROR:pywemo.ouimeaux_device:Unable to re-probe wemo at 10.22.22.1
INFO:pywemo.ouimeaux_device:Trying to reconnect with Hi Fi Systemline Sensor
INFO:pywemo.ouimeaux_device:Hi Fi Systemline Sensor Not found in try 0. Trying again in 0 seconds
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setup.xml HTTP/1.1" 200 4882
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setup.xml HTTP/1.1" 200 4882
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /setupservice.xml HTTP/1.1" 200 3278
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /timesyncservice.xml HTTP/1.1" 200 2190
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /eventservice.xml HTTP/1.1" 200 16357
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /deviceservice.xml HTTP/1.1" 200 3075
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /firmwareupdate.xml HTTP/1.1" 200 1826
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /rulesservice.xml HTTP/1.1" 200 11218
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /metainfoservice.xml HTTP/1.1" 200 1132
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /remoteaccess.xml HTTP/1.1" 200 3974
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /deviceinfoservice.xml HTTP/1.1" 200 10704
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /smartsetup.xml HTTP/1.1" 200 2539
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "GET /manufacture.xml HTTP/1.1" 200 642
INFO:pywemo.ouimeaux_device:Found Hi Fi Systemline Sensor again, updating local values
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "POST /upnp/control/WiFiSetup1 HTTP/1.1" 200 292
DEBUG:pywemo.ouimeaux_device:network status after 15.19 seconds: 1
DEBUG:urllib3.connectionpool:Starting new HTTP connection (1): 10.22.22.1:49152
DEBUG:urllib3.connectionpool:http://10.22.22.1:49152 "POST /upnp/control/WiFiSetup1 HTTP/1.1" 200 272
DEBUG:pywemo.ouimeaux_device:network status: 1
DEBUG:pywemo.ouimeaux_device:close status: success
DEBUG:pywemo.ouimeaux_device:SetSetupDoneStatus not able to be set (some devices do not have this method)
INFO:pywemo.ouimeaux_device:Wemo device connected to "Elsden-h", which took 15.93 total seconds over 1 connection attempt(s)
('1', 'success')
```

I'm basically happy to merge when you're happy. We can always make it better later.

Spectre5 committed:
Add information on reset and setup to README

Spectre5 committed:
minor README tweak

Spectre5 commented:
So in that log, it does report (at the end) that it successfully connected, right?  But in other tests you performed, did it actually end up reporting at the end that it did NOT connect (or get an exception raised)?

In this log, it appears that the device disconnected as you've found, before the first poll was performed.  Then the pywemo library has tried to re-connect to the lost device, probably at the first call to `status = wifisetup.GetNetworkStatus()['NetworkStatus']`.  My guess is that the device shut down its local AP and that is why the connection to the device was lost and it initially failed to connect.  Then your computer probably automatically reconnected to your home network and pywemo found the device connected on that network, via `_reconnect_with_device_by_discovery`.  When pywemo reconnected, it performed that first `GetNetworkStatus` and it was a 1 --- its already ON the home network at this point!  This is also why there is only one log message about the network status: `network status after 15.19 seconds: 1`.  So, in essence, this appears to me to happen pretty cleanly.  The `setup` method makes a call, pywemo does its thing and tries to connect to the device, it does reconnect successfully (on the home network), and then it reports at the end that it was successful (very last log message).

It is still surprising that it closed its local AP so quickly.  I wonder if some devices close it very quickly after getting the `ConnectHomeNetwork` command.  Although it must not be within 0.10 since the `ConnectHomeNetwork` gets sent twice with a 0.10 second delay between them.

If the device is "lost", then the timeout in the setup function is effectively ignored since pywemo will attempt to reconnect to it for a long period of time (multiple minutes before finally giving up and raising an ActionException via the `__call__` method).  But that is reasonable.

Spectre5 committed:
add better doc strings and status_delay argument

Spectre5 commented:
@pavoni, I've made some edits to address your review comments, mainly with the doc strings.  I've also added a section to the README regarding reset and setup and included a note about OpenSSL there.  Finally, I've added to the `setup` method a new argument named `status_delay` which allows the user to vary the delay between polling for the network status.

Spectre5 committed:
minor comment and log message clean up

Spectre5 committed:
reduce first status based on status_delay

Spectre5 committed:
dummy commit to re-trigger CI

Spectre5 committed:
undo dummy commit

Spectre5 committed:
doc string update

Spectre5 committed:
comment moved

Spectre5 commented:
ok, sorry about so many minor pushes!  Hopefully should be done, let me know when you take a look again.

Spectre5 committed:
1 letter changed, typo

Spectre5 committed:
check for bad pw and catch ActionException

Spectre5 committed:
re-connect -> reconnect

Spectre5 commented:
After looking into it more, I made two more changes:

- Catch `ActionException`, which is raised if a device is "lost" and pywemo cannot reconnect to it.  For example, on the line `status = wifisetup.GetNetworkStatus()['NetworkStatus']`, the action exception could be raised if the device is lost and it never is re-discovered.  So we can catch it to report that the setup status is uncertain since it might have finished connecting before it was lost.
- ~~It appears that `status ='2'` corresponds to a wrong password.  I decided to test a wrong password and I got this status repeatedly and it is the only time I've gotten that status, so I'm fairly certain it corresponds to a bad password.~~  So this status allows us to stop waiting for status changes and warn the user that the password was wrong.  I also made this a unique exception, `BadPassword`, so that an application using pywemo could easily catch that exception separately, which is the same reason for the `APNotFound` exception being unique.

Spectre5 committed:
add BadPassword exception

Spectre5 committed:
BadPassword -> ShortPassword

Spectre5 commented:
Looks like I mis-spoke in the previous message.  `status = '2'` is actually for a *short* password, not necessarily a wrong one.  Looks like Wemo requires a password of at least 8 characters as any password I tried with 7 or less characters returned this status code 2 immediately.  So I've changed that exception from `BadPassword` to `ShortPassword`.  A wrong password with 8 or more characters will just stay in `status = '0'`, so we can't be sure if the device didn't connect due to a wrong password or something else.

Spectre5 committed:
add status=3 handling

pavoni commented:
Now merged thanks so much for all the work.

Shall I release a new version - or do you want to wait until you have your other library ready to go?

Spectre5 commented:
I think go ahead and release.  I want to specifiy a minimum pywemo version in the other scripts requirements.

pavoni commented:
Now released as `0.5.4`.

Thanks for the all the work.

ViViDboarder commented:
@Spectre5 this is awesome! Great work here.
